esx_service/vmdk_ops.py

```diff
@@ -209,6 +213,91 @@ def make_create_cmd(opts, vmdk_path):
         return "{0} -d {1} -c {2} {3}".format(VMDK_CREATE_CMD, disk_format, size, vmdk_path)
 
 
+def cloneVMDK(vm_name, vmdk_path, opts={}, vm_uuid=None, vm_datastore=None):
+    logging.info("*** cloneVMDK: %s opts = %s", vmdk_path, opts)
+
+    # Get source volume path for cloning
+    error_info, tenant_uuid, tenant_name = auth.get_tenant(vm_uuid)
+    if error_info:
+        return err(error_info)
+
+    try:
+        src_volume, src_datastore = parse_vol_name(opts["clone-from"])
+    except ValidationError as ex:
+        return err(str(ex))
+    if not src_datastore:
+        src_datastore = vm_datastore
+    elif not vmdk_utils.validate_datastore(src_datastore):
+        return err("Invalid datastore '%s'.\n" \
+                    "Known datastores: %s.\n" \
+                    "Default datastore: %s" \
+                    % (datastore, ", ".join(get_datastore_names_list), src_datastore))
+    src_path, errMsg = get_vol_path(src_datastore, tenant_name)
+    if src_path is None:
+        return err("Failed to initialize source volume path {0}: {1}".format(src_path, errMsg))
+
+    src_vmdk_path = vmdk_utils.get_vmdk_path(src_path, src_volume)
+
+    # Verify if the source volume is in use.
+    attached, uuid, attach_as = getStatusAttached(src_vmdk_path)
+    if attached:
+        if handle_stale_attach(vmdk_path, kv_uuid):
+            return err("Source volume cannot be in use when cloning")
+
+    # We need to reauthorize with size info of the volume being cloned
+    if vm_uuid and vm_datastore:
+        src_vol_info = kv.get_vol_info(src_vmdk_path)
+        opts["size"] = src_vol_info["size"]
+        error_info, tenant_uuid, tenant_name = auth.authorize(vm_uuid, vm_datastore, "create", opts)
```

Comment:
Thanks for catching this @kerneltime, i understand that we're missing a explicit auth on the src_datastore. But its not only that, we're also missing a check of the src_volume file existence and a lock on it for the clone operation.

Thanks!